### PR TITLE
AutoMerge: Allow skipping sequential versions with author-approval criteria

### PR DESCRIPTION
**Pull request by ericphanson**
Based on #536 to avoid conflicts. That one should be merged first.

Requires https://github.com/JuliaRegistries/General/pull/98815

The flow here is:

* automerge comments, notes they can override
* package author comments "[merge approved]"
* GHA adds a label
* new label triggers automerge again
* automerge sees the new label and skips the sequential version check

FAQ:
* What happens if some one else comments "merge approved"? The label doesn't get applied, so nothing happens
* What happens if someone adds the label manually? The label does take effect. I.e. triage-level maintainers can facilitate author approval if needed. We need to trust them to not abuse this (I doubt it will happen and we can revoke triage access if folks aren't behaving)
* Does a "merge approved" comment block automerge unless it contains `[noblock]`? No, we special case it so comments that trigger author-approval count as noblock.
* Does AutoMerge's comment mentioning "merge approved" trigger the labelling workflow? No, it shouldn't, it is a bot user. (If it did, the label would still not be applied, since the registrator is not the package author)
* What happens if multiple guidelines use the author-approval mechanism, does the author have to indicate what they are approving? No, `merge approved` will apply to all of them
* What happens if the author comments before even seeing the automerge comment? They would end up approving whatever it says, but only 1 guideline is overridable at this point.
* What happens if the PR is from gitlab? We won't be able to find the author's github username in the PR text from Registrator, so the labeling workflow won't be able to check the commenter is the author, so the label won't get added. Triage-level maintainers can manually add the label in this case, but there isn't an automatic solution.
* What happens if the PR is not from Registrator? The label could still get added by the workflow (if it hits the regex to have the author name, and the person with that username comments), but automerge won't run, so the label won't do anything.

### CODE REVIEW

**Comment by DilumAluthge** [comment]
> * GHA adds a label
> * new label triggers automerge again

This might be a problem. If the label was added using the default `GITHUB_TOKEN` token, then IIRC the "label-added" event won't trigger any further GitHub Actions.

**Comment by DilumAluthge** [comment]
> > * GHA adds a label
> > * new label triggers automerge again
> 
> This might be a problem. If the label was added using the default `GITHUB_TOKEN` token, then IIRC the "label-added" event won't trigger any further GitHub Actions.

Can we make the "GHA adds a label" step use the custom `@JuliaTagBot` PAT we have? I believe that this would bypass the above issue.

Of course, maybe we should first verify that the above actually would be an issue.

**Comment by DilumAluthge** [comment]
Perhaps we should make the text specific to the guideline being approved? How about something like `/approve-merge-skip-version`?

Also, how about we add the `/` to the front, to emphasize that this is kind of a "command" that the package author is providing?

**Comment by DilumAluthge** [comment]
Even if we only get this working for GitHub-hosted packages at first, this still covers the majority of packages, so I still think it'll help reduce the manual registry-committer burden.

**Comment by ericphanson** [comment]
> This might be a problem. If the label was added using the default GITHUB_TOKEN token, then IIRC the "label-added" event won't trigger any further GitHub Actions.

Good point, I think that sounds right.

> Can we make the "GHA adds a label" step use the custom @JuliaTagBot PAT we have? I believe that this would bypass the above issue.
> Of course, maybe we should first verify that the above actually would be an issue.

Yeah, that makes sense. Since this can be done on the GHA by itself, IMO we can work on that once we merge down the stuff here and get RegistryCI updated on General. Then we can try it out and get the authentication working there.

> Perhaps we should make the text specific to the guideline being approved? How about something like /approve-merge-skip-version?
> Also, how about we add the / to the front, to emphasize that this is kind of a "command" that the package author is providing?

To me this seems overly complicated. Since RegistryCI prints all of the problems it found at once, the author can either approve merging or not knowing that information. I don't really feel like there's more than 1 decision the author actually has, since not approving 1 thing blocks merging (so ultimately they need to decide if they want to merge now or not).

I also think this will lead to more questions/issues where folks approve 1 thing but not another and are confused.

I think "merge approved" is straightforward and simpler.

> Even if we only get this working for GitHub-hosted packages at first, this still covers the majority of packages, so I still think it'll help reduce the manual registry-committer burden.

Agreed!

**Comment by ericphanson** [comment]
Re- the token, the [docs here](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow) seem clear that indeed `GITHUB_TOKEN` is insufficient for this, and we should use a PAT. We could use an existing one or make a new one for the task.

**Comment by DilumAluthge** [comment]
> We could use an existing one or make a new one for the task.

For now, probably easier to re-use the existing one (the `@JuliaTagBot` token that AutoMerge uses when merging a PR). But we might consider making a new one in the future.

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:466]:
```suggestion
    "write a comment that says `[merge approved]`.**")
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: src/AutoMerge/cron.jl
Comment (DilumAluthge) [reply src/AutoMerge/cron.jl:60]: As discussed in https://github.com/JuliaRegistries/General/pull/98815, we're going to compromise, and we'll require `[merge approved]`, but not require `[noblock]`.

File: src/AutoMerge/util.jl
Comment (ericphanson) [reply src/AutoMerge/util.jl:292]:
```suggestion
const AUTHOR_APPROVED_LABEL = "Override AutoMerge: package author approved"
```

File: test/automerge-unit.jl
Comment (ericphanson) [reply test/automerge-unit.jl:135]:
```suggestion
        @test AutoMerge.has_author_approved_label([GitHub.Label(; name="Override AutoMerge: package author approved")])
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [reply src/AutoMerge/guidelines.jl:466]: We will do https://github.com/JuliaRegistries/RegistryCI.jl/pull/539#discussion_r1468970719 instead.

File: test/automerge-unit.jl
Comment (ericphanson) [reply test/automerge-unit.jl:136]:
```suggestion
        @test AutoMerge.has_author_approved_label([GitHub.Label(; name="hi"), GitHub.Label(; name="Override AutoMerge: package author approved")])
```

File: src/AutoMerge/guidelines.jl
Comment (ericphanson) [reply src/AutoMerge/guidelines.jl:466]:
```suggestion
    "write a comment that contains the text: `[merge approved]`.**")
```

File: example_github_workflow_files/automerge.yml
Comment (ericphanson) [reply example_github_workflow_files/automerge.yml:17]:
```suggestion
    if: "${{  github.event.action != 'labeled' || (github.event.action == 'labeled' && (github.event.label.name == 'Override AutoMerge: name similarity is okay' || github.event.label.name == 'Override AutoMerge: package author approved')) }}"
```

File: test/automerge-unit.jl
Comment (ericphanson) [inline test/automerge-unit.jl:143]:
```suggestion
        @test !AutoMerge.pr_comment_is_blocking(GitHub.Comment(; body="[merge approved] abc"))
```

File: docs/src/guidelines.md
Comment (DilumAluthge) [inline docs/src/guidelines.md:26]:
```suggestion
        author_approved = false,
```

File: docs/src/guidelines.md
Comment (DilumAluthge) [inline docs/src/guidelines.md:56]:
```suggestion
        author_approved = false,
```

File: src/AutoMerge/cron.jl
Comment (DilumAluthge) [inline src/AutoMerge/cron.jl:60]:
```suggestion
    not_blocking = occursin("[noblock]", body(c)) || occursin("[merge approved]", lowercase(body(c)))
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

**Comment by DilumAluthge** [comment]
@ericphanson This needs a rebase, I think.

**Comment by ericphanson** [comment]
(updated)

File: example_github_workflow_files/automerge.yml
Comment (DilumAluthge) [inline example_github_workflow_files/automerge.yml:19]:
```suggestion
    if: "${{ github.event.action != 'labeled' || (github.event.action == 'labeled' && (github.event.label.name == 'Override AutoMerge: name similarity is okay' || github.event.label.name == 'Override AutoMerge: package author approved')) }}"
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: docs/src/guidelines.md
Comment (DilumAluthge) [inline docs/src/guidelines.md:26]:
```suggestion
        package_author_approved = false,
```

File: docs/src/guidelines.md
Comment (DilumAluthge) [inline docs/src/guidelines.md:56]:
```suggestion
        package_author_approved = false,
```

File: src/AutoMerge/cron.jl
Comment (DilumAluthge) [inline src/AutoMerge/cron.jl:59]:
```suggestion
    # Note: `[merge approved]` is not case sensitive, to match the semantics of `contains` on GitHub Actions
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:985]:
```suggestion
    package_author_approved::Bool # currently unused for new packages
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:1028]:
```suggestion
    package_author_approved::Bool,
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:1033]:
```suggestion
        (guideline_sequential_version_number, !this_pr_can_use_special_jll_exceptions && !package_author_approved),
```

File: src/AutoMerge/pull_requests.jl
Comment (DilumAluthge) [inline src/AutoMerge/pull_requests.jl:199]:
```suggestion
        package_author_approved=has_package_author_approved_label(data.pr.labels)
```

File: src/AutoMerge/util.jl
Comment (DilumAluthge) [inline src/AutoMerge/util.jl:292]:
```suggestion
const PACKAGE_AUTHOR_APPROVED_LABEL = "Override AutoMerge: package author approved"
```

File: src/AutoMerge/util.jl
Comment (DilumAluthge) [inline src/AutoMerge/util.jl:294]:
```suggestion
function has_package_author_approved_label(labels)
```

File: src/AutoMerge/util.jl
Comment (DilumAluthge) [inline src/AutoMerge/util.jl:298]:
```suggestion
        if label.name === PACKAGE_AUTHOR_APPROVED_LABEL
```

File: src/AutoMerge/util.jl
Comment (DilumAluthge) [inline src/AutoMerge/util.jl:300]:
```suggestion
            @debug "Found `$(PACKAGE_AUTHOR_APPROVED_LABEL)` label"
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:464]:
```suggestion
const PACKAGE_AUTHOR_APPROVAL_INSTRUCTIONS = string(
```

File: src/AutoMerge/guidelines.jl
Comment (DilumAluthge) [inline src/AutoMerge/guidelines.jl:478]:
```suggestion
        PACKAGE_AUTHOR_APPROVAL_INSTRUCTIONS,
```

File: test/automerge-unit.jl
Comment (DilumAluthge) [inline test/automerge-unit.jl:133]:
```suggestion
        @test !AutoMerge.has_package_author_approved_label(nothing)
```

File: test/automerge-unit.jl
Comment (DilumAluthge) [inline test/automerge-unit.jl:134]:
```suggestion
        @test !AutoMerge.has_package_author_approved_label([GitHub.Label(; name="hi")])
```

File: test/automerge-unit.jl
Comment (DilumAluthge) [inline test/automerge-unit.jl:135]:
```suggestion
        @test AutoMerge.has_package_author_approved_label([GitHub.Label(; name="Override AutoMerge: package author approved")])
```

File: test/automerge-unit.jl
Comment (DilumAluthge) [inline test/automerge-unit.jl:136]:
```suggestion
        @test AutoMerge.has_package_author_approved_label([GitHub.Label(; name="hi"), GitHub.Label(; name="Override AutoMerge: package author approved")])
```

**Pull request merged**
